src/Widgets/FirmwareUpdateRow.vala

```diff
@@ -0,0 +1,176 @@
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Marius Meisenzahl <mariusmeisenzahl@gmail.com>
+*/
+
+public class About.Widgets.FirmwareUpdateRow : Gtk.ListBoxRow {
+    public Device device { get; construct set; }
+
+    public signal void on_update_start ();
+    public signal void on_update_end ();
+
+    public FirmwareUpdateRow (Device device) {
+        Object (device: device);
+    }
+
+    construct {
+        var icon = new Gtk.Image.from_icon_name (device.icon, Gtk.IconSize.DND) {
+            pixel_size = 32
+        };
+
+        var device_name_label = new Gtk.Label (device.name) {
+            halign = Gtk.Align.START,
+            hexpand = true
+        };
+        device_name_label.get_style_context ().add_class (Granite.STYLE_CLASS_H3_LABEL);
+
+        var version_label = new Gtk.Label (device.latest_release.version) {
+            wrap = true,
+            xalign = 0
+        };
+
+        var grid = new Gtk.Grid () {
+            column_spacing = 12,
+            margin = 6
+        };
+        grid.attach (icon, 0, 0, 1, 2);
+        grid.attach (device_name_label, 1, 0);
+        grid.attach (version_label, 1, 1);
+
+        switch (device.latest_release.flag) {
+            case ReleaseFlag.IS_UPGRADE:
+                if (device.latest_release.version == device.version) {
+                    add_up_to_date_label (grid);
+                    break;
+                }
+
+                var update_button = new Gtk.Button.with_label (_("Update")) {
+                    valign = Gtk.Align.CENTER
+                };
+                update_button.clicked.connect (() => {
+                    on_update_start ();
+
+                    update.begin (device, device.latest_release, (obj, res) => {
+                        update.end (res);
+                        on_update_end ();
+                    });
+                });
+                grid.attach (update_button, 2, 0, 1, 2);
+                break;
+            default:
+                add_up_to_date_label (grid);
+                break;
+        }
+
+        add (grid);
+    }
+
+    private void add_up_to_date_label (Gtk.Grid grid) {
+        var update_to_date_label = new Gtk.Label (_("Up to date")) {
+            valign = Gtk.Align.CENTER
+        };
+        grid.attach (update_to_date_label, 2, 0, 1, 2);
+    }
+
+    private async void update (Device device, Release release) {
+        var path = yield FwupdManager.get_instance ().download_file (device, release.uri);
+
+        var details = yield FwupdManager.get_instance ().get_details (device, path);
+
+        if (details.caption != null) {
+            show_details_dialog (details);
```

Comment:
Would there be some expectation that we'd be able to cancel the upgrade in this dialog? We might not like some of the "details"? It looks like no matter what we do to this dialog, the upgrade happens anyway. Maybe that's an @elementary/ux question.

Reply:
Yeah that's a great point, there probably should be an option to cancel here